internal/controller/controllers/local_cluster_import_controller.go

```diff
@@ -0,0 +1,532 @@
+		}
+		clusterDeployment.Spec.ClusterInstallRef = &hivev1.ClusterInstallLocalReference{
+			Name:    r.localClusterName,
+			Group:   "extensions.hive.openshift.io",
+			Kind:    "AgentClusterInstall",
+			Version: "v1beta1",
+		}
+		clusterDeployment.Spec.Platform = hivev1.Platform{
+			AgentBareMetal: &agent.BareMetalPlatform{
+				AgentSelector: metav1.LabelSelector{
+					MatchLabels: map[string]string{"infraenv": "local-cluster"},
+				},
+			},
+		}
+		clusterDeployment.Spec.PullSecretRef = &v1.LocalObjectReference{
+			Name: pullSecret.Name,
+		}
+		clusterDeployment.Spec.ClusterName = r.localClusterName
+		clusterDeployment.Spec.BaseDomain = dns.Spec.BaseDomain
+		return nil
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, &clusterDeployment, mutateFn)
+	if err != nil {
+		return errors.Wrap(err, "could not create or ClusterDeployment")
+	}
+	return nil
+}
+
+func (r *LocalClusterImportReconciler) importLocalCluster(ctx context.Context, instance *aiv1beta1.AgentServiceConfig) error {
+
+	clusterVersion := &configv1.ClusterVersion{}
+	namespacedName := types.NamespacedName{
+		Name: "version",
+	}
+	err := r.client.Get(ctx, namespacedName, clusterVersion)
+	if err != nil {
+		return errors.Wrap(err, "unable to find cluster version")
+	}
+
+	if clusterVersion.Status.History != nil && clusterVersion.Status.History[0].State != configv1.CompletedUpdate {
+		return errors.New("the release image in the cluster version is not ready yet, please wait for installation to complete and then restart the assisted-service")
+	}
+
+	kubeConfigSecret, err := r.getSecret(ctx, "openshift-kube-apiserver", "node-kubeconfigs")
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch local cluster kubeconfigs")
+	}
+
+	pullSecret, err := r.getSecret(ctx, "openshift-machine-api", "pull-secret")
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch pull secret")
+	}
+
+	decodedJsonConfig, err := base64.StdEncoding.DecodeString(string(pullSecret.Data[".dockerconfigjson"]))
+	if err != nil {
+		return errors.Wrap(err, "unable to decode JSON config from pull secret")
+	}
+
+	dns := &configv1.DNS{}
+	namespacedName = types.NamespacedName{
+		Name: "cluster",
+	}
+	err = r.client.Get(ctx, namespacedName, dns)
+	if err != nil {
+		return errors.Wrap(err, "could not fetch DNS")
+	}
+
+	proxy := &configv1.Proxy{}
+	namespacedName = types.NamespacedName{
+		Name: "cluster",
+	}
+	err = r.client.Get(ctx, namespacedName, proxy)
+	if err != nil {
+		return errors.Wrap(err, "could not fetch proxy")
+	}
+
+	numberOfControlPlaneNodes := 0
+	nodeList := &v1.NodeList{}
+	err = r.client.List(ctx, nodeList)
+	if err != nil {
+		return errors.Wrap(err, "error while fetching nodes")
+	}
+	for _, node := range nodeList.Items {
+		for nodeLabelKey := range node.Labels {
+			if nodeLabelKey == "node-role.kubernetes.io/control-plane" {
+				numberOfControlPlaneNodes++
+			}
+		}
+	}
+
+	if clusterVersion.Status.History[0].Image == "" {
```

Comment:
I think it would be better to use `Desired` here so you don't have to rely on ordering.

That's what we do [elsewhere](https://github.com/openshift/assisted-service/blob/7080d0d41e3795bd8c36dea49a25bdfc061938df/internal/controller/controllers/preprovisioningimage_controller.go#L514) when we need the current release image.